src/index.ts

```diff
@@ -137,54 +139,71 @@ class CruxPayPeer {
         // TODO: Need to validate options
 
-        this._getEncryptionKey = _options.getEncryptionKey;
+        this._getEncryptionKey = this._options.getEncryptionKey;
         // log.debug(`Encryption key:`, this._getEncryptionKey())
 
         // Setting up the default modules as fallbacks
         this._storage =  this._options.storage || new storage.LocalStorage();
         this._encryption = this._options.encryption || encryption.Encryption;
         this._nameService = this._options.nameService;
+        this._keyPair = this._options.keyPair;
         this.walletClientName = this._options.walletClientName;
 
         log.info(`Config mode:`, config.CONFIG_MODE);
         log.info(`CruxPayPeer Initialised`);
     }
 
     public async init() {
-        let configService;
-
+        let configService: BlockstackConfigurationService;
         if (this._hasPayIDClaimStored()) {
-            const payIDClaim = this._storage.getJSON("payIDClaim");
+            log.debug("using the stored payIDClaim");
+            const payIDClaim = (this._storage.getJSON("payIDClaim") as ICruxPayClaim);
             // log.debug(`Local payIDClaim:`, payIDClaim)
-            this._setPayIDClaim(new PayIDClaim(payIDClaim as ICruxPayClaim, { getEncryptionKey: this._getEncryptionKey }));
-            const ns: blockstackService.BlockstackService = new blockstackService.BlockstackService({domain: this.walletClientName + CRUX_DOMAIN_SUFFIX});
-            await (this._payIDClaim as PayIDClaim).decrypt();
-            await ns.restoreIdentity((this._payIDClaim as PayIDClaim).virtualAddress as string, {secrets: (this._payIDClaim as PayIDClaim).identitySecrets});
-            const status = await ns.getRegistrationStatus({secrets: (this._payIDClaim as PayIDClaim).identitySecrets});
-            await (this._payIDClaim as PayIDClaim).encrypt();
-            if (status.status === blockstackService.SubdomainRegistrationStatus.DONE) {
-                configService = new BlockstackConfigurationService(this.walletClientName, (this._payIDClaim as PayIDClaim).virtualAddress);
-            } else {
-                configService = new BlockstackConfigurationService(this.walletClientName);
+
+            // if a keyPair is provided, add a validation check on the cruxID stored
+            if (this._keyPair) {
+                const CruxIDArray = await getCruxIDByAddress(config.BLOCKSTACK.BNS_NODES[0], this.walletClientName, this._keyPair.address);
+                if (!CruxIDArray.includes(payIDClaim.virtualAddress as string)) {
+                    throw errors.ErrorHelper.getPackageError(errors.PackageErrorCode.KeyPairMismatch);
+                }
             }
+            this._setPayIDClaim(new PayIDClaim(payIDClaim, { getEncryptionKey: this._getEncryptionKey }));
+            configService = await this._getConfigService();
             await this._initializeNameService(configService);
             await this._restoreIdentity();
```

Comment:
looking from existing code.. there are two _restoreIdentity and nameService. restoreIdentity... 
_restoreIdentity internally calls (nameService. restoreIdentity) and init is first calling  `nameService. restoreIdentity` and `_restoreIdentity` afterwards. Not sure if I am able to follow the reason here. Why calls can't be combined?

Reply:
One is done using a temporary NameService instance (in the Init) only to get the registration status of the cruxId. A successful registration would provide config overrides which will lead to the final NameService initialisation.